src/search.c

```diff
@@ -153,7 +150,78 @@ static void free_bundle_result_data(void *data)
 	free(br);
 }
 
-static void print_final_results(void)
+/* unmark all bundle_results in bl as seen */
+static void unsee_bundle_results(struct list *bl)
+{
+	struct bundle_result *br;
+	struct list *ptr = list_head(bl);
+
+	while (ptr) {
+		br = ptr->data;
+		br->seen = false;
+		ptr = ptr->next;
+	}
+}
+
+/* recursively calculate size from a complete list of bundle_result structs */
+static long calculate_size(char *bname, struct list *bundle_info, bool installed)
+{
+	struct list *ptr;
+	struct bundle_result *bi;
+	long size = 0;
+	ptr = list_head(bundle_info);
+	while (ptr) {
+		bi = ptr->data;
+		ptr = ptr->next;
+		if (bi->seen || strcmp(bname, bi->bundle_name) != 0) {
+			continue;
+		}
+
+		/* this was a top-level match or an include match
+		 * add the contentsize if we are computing for an already-installed bundle
+		 * or if the bundle is not already included */
+		if (installed || !is_tracked_bundle(bname)) {
+			size += bi->topsize;
+			bi->seen = true;
+		}
+
+		/* add bundle sizes for includes as well */
+		struct list *ptr2;
+		ptr2 = list_head(bi->includes);
+		while (ptr2) {
+			char *inc = NULL;
+			inc = ptr2->data;
+			ptr2 = ptr2->next;
+			/* recursively add included sizes */
+			size += calculate_size(inc, bundle_info, installed);
+		}
+	}
+
+	return size;
+}
+
+static void apply_size_penalty(struct list *bundle_info)
+{
+	struct bundle_result *b;
+	struct list *ptr;
+
+	ptr = list_head(results);
+
+	while (ptr) {
+		b = ptr->data;
+		ptr = ptr->next;
+
+		/* unsee all bundles before calculating size */
+		unsee_bundle_results(bundle_info);
+		/* calculate bundle size for recursive includes */
+		b->size = calculate_size(b->bundle_name, bundle_info, b->is_tracked);
```

Comment:
The benefits: no need to explain the "installed" boolean (that section was confusing because the dual role of 'size'); only do the recursive walk once (instead of one walk per bundle).

Reply:
Sorry, I'm not really following your suggestion. Do you mean making calculate_size perform the calculations on the results list itself? I'm sure I'm misunderstanding but I don't see the benefit of moving the for loop one function down.

Reply:
If I understand correctly, right now you use calculate_size to perform two types of calculations: (a) the installed size (if the toplevel bundle is tracked) or (b) the size to be installed (if not). Because the result for (a) could be different than result from (b) for each bundle, we need to unsee all the bundles before making the next calculation. This cause you to do the "bundle walk" for each bundle.

The suggestion is to change bundle_info to keep score of both "installed_size" and "total_size" (and anything else that is useful), and perform the bundle walk for all unseen bundles, until there is no bundle unseen. We don't care about the return value of calculate_size anymore, as all the data would be in the bundle_info. Once those numbers are in, you can set "b->size" based on whether the bundle is tracked or not.

Right now bundle walk is O(n^2) and doing it for each bundle is O(n^3). For small n we are fine, but with eyes on a large number of bundles in the future, this may make a difference. I also argue that will make the code a bit easier to understand: no need to explain the install/will-be-installed logic -- so I suggested it regardless the performance.

Reply:
So, I tried this and it started to get rather complicated due to the includes list just being a list of char * and not a list of bundle_result or manifest *. At some point recursion based on string matching has to occur which makes this suggestion messy, at least as I try to implement it.

A better solution I think for the package-level bundles idea is to make includes a flat list of all nested includes. This way recursion is resolved on the server side where performance isn't nearly as important as client-side. So if bundleA includes bundleB which includes bundleC the list of includes for bundleA will be:
```
includes:     bundleB
includes:     bundleC
includes:     os-core
```

Then on the client side no recursion is necessary to resolve the dependency tree. I think at that point it makes sense to revisit this function and make it much simpler.

Reply:
> At some point recursion based on string matching has to occur which makes this suggestion messy, at least as I try to implement it.

I'm fine if we improve on this iteratively. In the server side when we did this kind of operation we had maps to make things less messier. Kind of miss them sometimes in C :-)

> A better solution I think for the package-level bundles idea is to make includes a flat list of all nested includes.

IMHO, I would rather not moving towards this unless we have more data about client no handling the computation. One of the issue is if the upstream has bundleA, that includes bundleB, that includes bundleC. If we (as a mix) bundleB in a way that adds or removes other bundles, we need to modify bundleA as well. This kind of cripples any attempt to use bundles as is, as we need to rewrite its manifest.

Reply:
> we need to modify bundleA

That's a good point for people making mixes, hadn't thought about that.